api/src/main/java/org/openmrs/module/pihcore/htmlformentry/action/CloseCovidProgramAction.java

```diff
@@ -0,0 +1,112 @@
+
+    private static String DISPOSITION_DISCHARGE = "discharge";
+
+    private static String CONCEPT_TRANSFER_OUT = "PATIENT TRANSFERRED OUT";
+
+    private static String CONCEPT_DIED = "PATIENT DIED";
+
+    private static String CONCEPT_DISCHARGE = "TREATMENT STOPPED";
+
+    @Override
+    public void applyAction(FormEntrySession formEntrySession) {
+        closeCovidProgram(formEntrySession, Context.getProgramWorkflowService(), Context.getService(DispositionService.class), Context.getConceptService());
+    }
+
+
+    private void closeCovidProgram(FormEntrySession formEntrySession, ProgramWorkflowService programWorkflowService, DispositionService dispositionService, ConceptService conceptService) {
+        Patient patient = formEntrySession.getPatient();
+        Program covid = programWorkflowService.getProgramByUuid(COVID_PROGRAM_UUID);
+        Encounter encounter = formEntrySession.getEncounter();
+        // use the encounter date *without* the time component
+        Date encounterDate = (new DateTime(encounter.getEncounterDatetime())).withTimeAtStartOfDay().toDate();
+
+        // we will update any COVID program started on or before the encounter date AND EITHER closed on or after encounter date OR not closed at all
+        // note that this should handle moving a completion date *earlier* but *not later*:
+        // ie if a patient is enrolled on Jan 11 and program is completed on Jan 14 and discharge form is entered on Jan 12, we assume
+        // Jan 12 is the correct completion date and update
+        // however, if a discharge form is entered on Jan 16, the program from Jan 11 to Jan 14 is *not* updated
+
+        List<PatientProgram> candidates = programWorkflowService.getPatientPrograms(patient, covid, null,
+                encounterDate, encounterDate, null, false);
+
+        if (candidates != null) {
+
+            if (candidates.size() > 1) {
+                log.warn("More than one COVID program enrolment for patient " + patient.getId() + " on date " + encounterDate);
+            }
+
+            for (PatientProgram patientProgram : candidates) {
+
+                // set the date completed to the encounter date
+                patientProgram.setDateCompleted(encounterDate);
+
+                // find the disposition
+                DispositionDescriptor dispositionDescriptor = dispositionService.getDispositionDescriptor();
+
+                Disposition disposition = null;
+
+                for (Obs candidate : encounter.getObsAtTopLevel(false)) {
+                    if (dispositionDescriptor.isDisposition(candidate)) {
+                        disposition = dispositionService.getDispositionFromObsGroup(candidate);
+                        break;
+                    }
+                }
+
+                // set program outcome based on dispositoion
+                if (disposition != null) {
+                    if (disposition.getUuid().equals(DISPOSITION_DEATH)) {
```

Comment:
I wouldn't expect `disposition.getUuid()` to ever equal `"markPatientDead"`?

Reply:
That's a good expectation, but it does... I don't know why that field is called uuid, it probably should be called key.  I should comment, however...

https://github.com/PIH/openmrs-config-zl/blob/master/configuration/pih/pih-dispositions-haiti.json#L50

Reply:
:open_mouth: